python/gym_jiminy/toolbox/gym_jiminy/toolbox/math/generic.py

```diff
@@ -74,3 +74,116 @@ def quat_to_rpy(quat: np.ndarray) -> np.ndarray:
     if quat.ndim == 1:
         return rpy[0]
     return rpy
+
+
+@nb.jit(nopython=True, nogil=True, inline='always')
+def quat_to_matrix(quat: np.ndarray) -> np.ndarray:
+    """Compute the Rotation Matrix representation of a single or a
+    batch of quaternions.
+
+    :param quat: 1D array, a 2D array or nD array whose first dimensions
+                 corresponds to the number of individual quaternions, and the
+                 last to the 4 coordinates [qx, qy, qz, qw].
+    """
+    qx, qy, qz, qw = np.atleast_2d(quat)[..., -4:].T
+    col_1 = np.stack(
+        (
+            qx * qx - qy * qy - qz * qz + qw * qw,
+            2 * (qx * qy + qz * qw),
+            2 * (qx * qz - qy * qw),
+        ),
+        -1,
+    )
+    col_2 = np.stack(
+        (
+            2 * (qx * qy - qz * qw),
+            -qx * qx + qy * qy - qz * qz + qw * qw,
+            2 * (qy * qz + qx * qw),
+        ),
+        -1,
+    )
+    col_3 = np.stack(
+        (
+            2 * (qx * qz + qy * qw),
+            2 * (qy * qz - qx * qw),
+            -qx * qx - qy * qy + qz * qz + qw * qw,
+        ),
+        -1,
+    )
+    matrix = np.stack((col_1, col_2, col_3), -1)
+    if quat.ndim == 1:
+        return matrix[0, :, :]
+    return matrix
+
+
+@nb.jit(nopython=True, nogil=True, inline='always')
+def matrix_to_quat(matrix: np.ndarray) -> np.ndarray:
+    """Compute the [qx, qy, qz, qw] Quaternion representation of a single or a
+    batch of rotation matrices.
+
+    :param matrix: 2D array, a 3D array or a nD array whose first dimensions
+                   corresponds to the number of individual rotation matrices,
+                   and the two last to the 3-by-3 rotation matrices.
+    """
+    quat = np.stack(
+        (
+            matrix[..., 2, 1] - matrix[..., 1, 2],
+            matrix[..., 0, 2] - matrix[..., 2, 0],
+            matrix[..., 1, 0] - matrix[..., 0, 1],
+            1 + matrix[..., 0, 0] + matrix[..., 1, 1] + matrix[..., 2, 2],
+        ),
+        -1,
+    )
+    norm = np.sqrt(np.sum(np.atleast_2d(quat * quat), -1))
```

Comment:
Just add a warning stating that you assume unit rotation matrix, so it should be already normalized.

Eventually, you could avoid rounding errors by using fast first-order normalization:
```python
    quat *= (3.0 - np.sum(quat * quat, 0)) / 2.0
```